File: SQL/Select/SQLSelectOrderByField.vb

' ___________________________________________________
'
'  © Hi-Integrity Systems 2010. All rights reserved.
'  www.hisystems.com.au - Toby Wicks
' ___________________________________________________
'

Option Strict On
Option Explicit On

Namespace SQL

    Public Class SQLSelectOrderByField

        Private pstrName As String
        Private pobjTable As SQLSelectTable
        Private peOrder As OrderBy = OrderBy.Ascending
        Private peAggregateFunction As SQL.AggregateFunction = Global.DatabaseObjects.SQL.AggregateFunction.None

        Friend Sub New()

            MyBase.New()

        End Sub

        Public Property Name() As String
            Get

                Return pstrName

            End Get

            Set(ByVal Value As String)

                pstrName = Value.Trim

            End Set
        End Property

        Public Property Order() As OrderBy
            Get

                Return peOrder

            End Get

            Set(ByVal Value As OrderBy)

                peOrder = Value

            End Set
        End Property

        Public Sub OrderingReverse()

            If peOrder <> OrderBy.None Then
                peOrder = Not peOrder
            End If

        End Sub

        Public Property AggregateFunction() As SQL.AggregateFunction
            Get

                Return peAggregateFunction

            End Get

            Set(ByVal Value As SQL.AggregateFunction)

                peAggregateFunction = Value

            End Set
        End Property

        Public Property Table() As SQLSelectTable
            Get

                Return pobjTable

            End Get

            Set(ByVal Value As SQLSelectTable)

                pobjTable = Value

            End Set
        End Property

        Friend ReadOnly Property SQL(ByVal eConnectionType As Database.ConnectionType) As String
            Get

                Dim strSQL As String = String.Empty

                If Me.Name = String.Empty Then
                    Throw New Exceptions.DatabaseObjectsException("Order By field has not been set.")
                End If

                If Me.AggregateFunction > 0 Then
                    strSQL = SQLConvertAggregate(Me.AggregateFunction) & "("
                End If

                strSQL &= SQLFieldNameAndTablePrefix(Me.Table, Me.Name, eConnectionType)

                If Me.AggregateFunction > 0 Then
                    strSQL &= ")"
                End If

                Select Case Me.Order
                    Case OrderBy.Ascending
                    Case OrderBy.Descending
                        strSQL &= " DESC"
                End Select

                Return strSQL

            End Get
        End Property

    End Class

End Namespace

File: SQL/Select/SQLSelectOrderByField.cs
// _________________________________________________________________________
//
//  © Hi-Integrity Systems 2010. All rights reserved.
//  www.hisystems.com.au - Toby Wicks
// 
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//	    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// _________________________________________________________________________
//

using System.Collections;
using System;
using System.Data;

namespace DatabaseObjects.SQL
{
	public class SQLSelectOrderByField
	{
		private string pstrName;
		private SQLSelectTable pobjTable;
		private OrderBy peOrder = OrderBy.Ascending;
		private AggregateFunction peAggregateFunction = AggregateFunction.None;
			
		internal SQLSelectOrderByField()
		{
		}
			
		public string Name
		{
			get
			{
				return pstrName;
			}
				
			set
			{
                pstrName = value;
			}
		}
			
		public OrderBy Order
		{
			get
			{
				return peOrder;
			}
				
			set
			{
				peOrder = value;
			}
		}
			
		public void OrderingReverse()
		{
			if (peOrder != OrderBy.None)
				peOrder = ~peOrder;
		}
			
		public SQL.AggregateFunction AggregateFunction
		{
			get
			{
				return peAggregateFunction;
			}
				
			set
			{
				peAggregateFunction = value;
			}
		}
			
		public SQLSelectTable Table
		{
			get
			{
				return pobjTable;
			}
				
			set
			{
				pobjTable = value;
			}
		}
	}
}
